src/drivers/net/marvell/aqc1xx.c

```diff
@@ -0,0 +1,634 @@
+ *   PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR HOLDER OR 
+ *   CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, 
+ *   EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, 
+ *   PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; 
+ *   OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, 
+ *   WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) 
+ *   ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE 
+ *   POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+FILE_LICENCE ( BSD2 );
+
+#include <stdint.h>
+#include <stdio.h>
+#include <string.h>
+#include <unistd.h>
+#include <errno.h>
+#include <byteswap.h>
+#include <ipxe/netdevice.h>
+#include <ipxe/ethernet.h>
+#include <ipxe/if_ether.h>
+#include <ipxe/iobuf.h>
+#include <ipxe/malloc.h>
+#include <ipxe/pci.h>
+#include <ipxe/profile.h>
+#include <compiler.h>
+
+#include "aqc1xx.h"
+
+
+static int atl_ring_alloc(const struct atl_nic *nic, struct atl_ring *ring,
+			  uint32_t desc_size, uint32_t reg_base)
+{
+	physaddr_t phy_addr;
+
+	/* Allocate ring buffer.*/
+	ring->length = ATL_RING_SIZE * desc_size;
+	ring->ring = malloc_phys(ring->length, ATL_RING_ALIGN);
+	if (!ring->ring)
+		goto err_alloc;
+	ring->sw_head = ring->sw_tail = 0;
+
+	memset(ring->ring, 0, ring->length);
+
+	/* Write ring address (hi & low parts).*/
+	phy_addr = virt_to_bus(ring->ring);
+	ATL_WRITE_REG((uint32_t)phy_addr, reg_base + 0);
+	ATL_WRITE_REG((uint32_t)(((uint64_t)phy_addr) >> 32), reg_base + 4);
+	/* Write ring length.*/
+	ATL_WRITE_REG(ATL_RING_SIZE, reg_base + 8);
+
+	/* #todo: reset head and tail pointers */
+	DBG("ATLANTIC %p ring is at [%08llx,%08llx), reg base %#x\n",
+	    nic, ((unsigned long long)phy_addr),
+	    ((unsigned long long) phy_addr + ring->length), reg_base);
+
+	return 0;
+
+err_alloc:
+	if (ring->ring)
+		free_phys(ring->ring, ring->length);
+
+	ring->ring = NULL;
+	ring->length = 0;
+
+	return -ENOMEM;
+}
+
+static void atl_ring_free(struct atl_ring *ring)
+{
+	if (ring->ring)
+		free_phys(ring->ring, ring->length);
+	ring->ring = NULL;
+	ring->length = 0;
+}
+
+static void atl_ring_next_dx(unsigned int *val)
+{
+	++(*val);
+	if (*val == ATL_RING_SIZE)
+		*val = 0;
+}
+
+int atl_ring_full(const struct atl_ring *ring)
+{
+	unsigned int tail = ring->sw_tail;
+	atl_ring_next_dx(&tail);
+	return tail == ring->sw_head;
+}
+
+void atl_rx_ring_fill(struct atl_nic *nic)
+{
+	struct atl_desc_rx *rx;
+	struct io_buffer *iobuf;
+	physaddr_t address;
+	unsigned int refilled = 0;
+
+	/* Refill ring */
+	while (!atl_ring_full(&nic->rx_ring)) {
+
+		/* Allocate I/O buffer */
+		iobuf = alloc_iob(ATL_RX_MAX_LEN);
+		if (!iobuf) {
+			/* Wait for next refill */
+			break;
+		}
+
+		/* Get next receive descriptor */
+		rx = (struct atl_desc_rx *)nic->rx_ring.ring +
+		nic->rx_ring.sw_tail;
+
+		/* Populate receive descriptor */
+		address = virt_to_bus(iobuf->data);
```

Comment:
Please use `alloc_rx_iob()` instead so that your driver will work on systems with an IOMMU enabled.  You can use `drivers/net/intel.c` as a reference for how to do this.